### PR TITLE
Hotfix for jsonld context

### LINKED ISSUE #548: Bug in marshalling jsonld objects
There is a bug during marshalling of jsonld objects. When dropping object to yaml in function asjsonld (with retain_collection_types=True), collection types are not getting sourced into jsonld context correctly.

The accepted solution would be that those fields within a collection get put into a top level context object.

related: #547

### PR DESCRIPTION
**Pull request by Panaetius**
Makes `@context` in KG nested and type specific, solving issues of name collisions and missing triples.

See https://www.w3.org/TR/json-ld11/#scoped-contexts for details on scoping contexts.

In-Depth description of problem&solution:
E.g. `Generation.entity` can refer to different python classes which can occur side by side. In order to process the KG, we need to combine all JSON values in the KG with their semantic type (coming from an Ontology). We have to deal with ambiguities, like `something.project` referring to `schema:project` in one place and `somethingelse.project` referring to `foaf:project` somewhere else. To deal with this, we need to use local, scoped contexts that are only applicable for a single class, not some global context that's applicable to everything. We achieve vertical scoping (A child class having the same property as parent but with different semantics) by nesting contexts following the same structure as the value graph. We can achieve horizontal scoping (A class in a collection having the same property name as a sibling but with different semantics) by scoping the nested contexts on @type of each class. Since scoping  on types only works for single types, but most of our classes have more than one semantic type, we need to repeat the context for each semantic type of a class, making the context a bit of a mess and unwieldy, but allowing consistent and robust usage of the context everywhere.

This PR is not based on master (!), but on PR #765 as it is needed to test that PR and I already started working on it there when I though there was a quick fix possible.

Closes #548

### CODE REVIEW

File: renku/core/models/jsonld.py [hunk 1]
```diff
@@ -185,13 +214,15 @@ def wrap(cls):
...
+                raise TypeError(
+                    'Type {0!r} is already registered for class {1!r}.'.format(
+                        jsonld_cls._jsonld_type,
+                        jsonld_cls.__type_registry__[type_],
+                    )
                 )
-            )
-        jsonld_cls.__type_registry__[type_] = jsonld_cls
+
+            jsonld_cls.__type_registry__[type_] = jsonld_cls
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/models/jsonld.py]: Good catch, I accidentally indented that line too much when changing the exception handling.

File: renku/core/models/jsonld.py [hunk 2]
```diff
@@ -136,18 +138,42 @@ def wrap(cls):
...
                         current_context = {'@id': ctx}
                     elif not isinstance(current_context, dict):
                         current_context = {'@id': current_context}
 
                     current_context['@context'] = merge_ctx
+                else:
+                    scoped_properties.append(key)
+
+                    for subcls in all_subclasses:
+                        if hasattr(subcls, '_jsonld_context'):
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/models/jsonld.py]: Yes, you're right.

I'm redoing this part anyways, it's not really `all_subclasses` as I gave up on the idea of automatically detecting all subclasses of a class, so the naming is all wrong and the code can be simplified.

**Comment by rokroskar** [comment]
would it be easier to review if the PR was made to the branch of #765?

File: renku/core/models/jsonld.py [hunk 3]
```diff
@@ -114,57 +115,32 @@ def wrap(cls):
                     types.append(subtype)
 
             for key, value in getattr(subcls, '_jsonld_context', {}).items():
-                if key in context and context[key] != value:
+                if key in ctx and ctx[key] != value:
```
Comment (rokroskar) [inline renku/core/models/jsonld.py]: why rename? I find it confusing because `ctx` is usually used for click context

**Pull request merged**